### PR TITLE
*: add pdctl

### PR DESCRIPTION
**Pull request by nolouch**
now simply support config Get and Set
PTAL @huachaohuang @siddontang @overvenus

### CODE REVIEW

File: cmd/pdctl/main.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package main
```
Comment (siddontang) [inline cmd/pdctl/main.go]: use pd-ctl as the folder name

File: Makefile [hunk 2]
```diff
@@ -13,7 +13,7 @@ LDFLAGS += -X "github.com/pingcap/pd/server.PDGitHash=$(shell git rev-parse HEAD
 # See more: https://godoc.org/path/filepath#Match
 COVERIGNORE := "cmd/pd-server/main.go,server/api/bindata_assetfs.go"
 
-default: build
+default: build ctl
```
Comment (siddontang) [inline Makefile]: put  `ctl` in `build` directly. 

`make build` can build `pd-server` and `pd-ctl`.

File: cmd/pd-ctl/main.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+var url string
+
+func init() {
+	flag.StringVar(&url, "u", "http://127.0.0.1:2379", "the pd address")
+}
+
+func main() {
+	flag.Parse()
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, syscall.SIGINT)
```
Comment (siddontang) [inline cmd/pd-ctl/main.go]: capture other SIGNALs like pd-server too.

File: Makefile [hunk 4]
```diff
@@ -24,11 +24,16 @@ build-fe:
 	go get github.com/elazarl/go-bindata-assetfs/...
 	cd server/api && go-bindata-assetfs -pkg api templates/... && cd -
 
-build:
+build: ctl
```
Comment (siddontang) [inline Makefile]:
```
build:
        rm -rf vendor && ln -s _vendor/vendor vendor
  	$(GO) build -ldflags '$(LDFLAGS)' -o bin/pd-server cmd/pd-server/main.go
        $(GO) build  -o bin/pd-ctl cmd/pd-ctl/main.go
  	rm -rf vendor
```

File: pdctl/command/exit_command.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 The etcd Authors
```
Comment (siddontang) [inline pdctl/command/exit_command.go]: why using etcd doc.go directly?

File: pdctl/command/config_command.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	url := getAddressFromCmd(cmd, schedulePrefix)
+	var value interface{}
+	data := make(map[string]interface{})
+
+	r, err := http.Get(url)
+	if err != nil {
+		fmt.Printf("Failed to set config:[%s]\n", err)
+		return
+	}
```
Comment (siddontang) [inline pdctl/command/config_command.go]: defer r.Body.Close()

File: pdctl/command/config_command.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	if r.StatusCode != http.StatusOK {
+		fmt.Println(r.StatusCode, "error")
```
Comment (siddontang) [inline pdctl/command/config_command.go]: r.Body.Close()

File: pdctl/command/config_command.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	url = getAddressFromCmd(cmd, configPrefix)
+	r, err = http.Post(url, "application/json", bytes.NewBuffer(req))
+	if err != nil {
+		fmt.Printf("Failed to set config:[%s]\n", err)
+	}
+	defer r.Body.Close()
+	if r.StatusCode == http.StatusOK {
+		fmt.Println("Success!")
+	} else {
+		fmt.Println("Failed:")
```
Comment (siddontang) [inline pdctl/command/config_command.go]: failed with error code

**Comment by siddontang** [comment]
Rest LGTM

File: Makefile [hunk 9]
```diff
@@ -13,7 +13,7 @@ LDFLAGS += -X "github.com/pingcap/pd/server.PDGitHash=$(shell git rev-parse HEAD
 # See more: https://godoc.org/path/filepath#Match
 COVERIGNORE := "cmd/pd-server/main.go,server/api/bindata_assetfs.go"
 
-default: build
+default: build
```
Comment (overvenus) [inline Makefile]: Remove trailing space.

File: pdctl/command/config_command.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	conf.AddCommand(NewShowConfigCommand())
+	conf.AddCommand(NewSetConfigCommand())
+	return conf
+}
+
+// NewShowConfigCommand return a show subcommand of configCmd
+func NewShowConfigCommand() *cobra.Command {
+	sc := &cobra.Command{
+		Use:   "show",
+		Short: "show config fo PD",
```
Comment (overvenus) [inline pdctl/command/config_command.go]: Typo `fo`?

File: pdctl/command/config_command.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+		Run:   configShowCommandF,
```
Comment (overvenus) [inline pdctl/command/config_command.go]: > `conf.AddCommand(NewShowConfigCommand())`

So why not name it `showConfigCommandF`？

File: pdctl/command/config_command.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+		return
+	}
+	defer r.Body.Close()
+	if r.StatusCode != http.StatusOK {
+		fmt.Println(r.StatusCode, "error")
+		return
+	}
+
+	io.Copy(os.Stdout, r.Body)
+
```
Comment (overvenus) [inline pdctl/command/config_command.go]: Remove the empty line.

File: pdctl/command/config_command.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+		fmt.Printf("Failed to get config:[%s]\n", err)
+		return
+	}
+	defer r.Body.Close()
+	if r.StatusCode != http.StatusOK {
+		fmt.Println(r.StatusCode, "error")
+		return
+	}
+
+	io.Copy(os.Stdout, r.Body)
```
Comment (overvenus) [inline pdctl/command/config_command.go]: How about `fmt.Fprintf()`? Don't forget to check errors.

File: server/api/config_test.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	//	"fmt"
```
Comment (overvenus) [inline server/api/config_test.go]: Remove it.

File: pdctl/command/config_command.go (hunk 11 above)
Comment (nolouch) [inline pdctl/command/config_command.go]: here not a subcommand , it is a action

Comment (overvenus) [inline pdctl/command/config_command.go]: An action should start with a verb.

File: pdctl/command/config_command.go (hunk 13 above)
Comment (nolouch) [inline pdctl/command/config_command.go]: r.Body is an io.Reader . here use io.Copy is ok. I have checked the response statusCode.

Comment (overvenus) [inline pdctl/command/config_command.go]: Then please check error of Copy.

> `func Copy(dst Writer, src Reader) (written int64, err error)`

Comment (nolouch) [inline pdctl/command/config_command.go]: no need to check it in here ,it  just likes print. and whatever error happened,it will return.

**Comment by nolouch** [comment]
PTAL @huachaohuang

**Comment by huachaohuang** [comment]
LGTM

**Pull request merged**